### PR TITLE
Abstract evaluation of switch cases (simple).

### PR DESCRIPTION
**Pull request by neilmacintosh**
Adds abstract evaluation for switch cases. Effectively treats the switch case as a series of if-else statements to achieve the same outcome.

Currently does not support case blocks that result in an abrupt completion through `return/throw/continue`. That is work-to-go in the future.

A future PR will also expand on the tests to provide more coverage (and look for corner cases).

Test plan: Added more tests to serializer/abstract/Switch.js

### CODE REVIEW

File: src/evaluators/SwitchStatement.js [hunk 1]
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
...
+          throw ReportNoPossiblyNormalCompletionError();
+        }
+
+        result = GetValue(r, result, realm);
+      }
+      caseIndex++;
+    }
+
+    // TODO work out what to do with Reference
+    if (result instanceof Reference) {
```
Comment (hermanventer) [inline src/evaluators/SwitchStatement.js]: Can't happen.

File: src/evaluators/SwitchStatement.js [hunk 2]
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
...
+
+        // TODO understand Reference and how to deal with it better
+        // TODO come up with a strategy for handling PossiblyNormalCompletion correctly
+        if (r instanceof Reference) {
+          throw ReportNoReferenceError();
+        } else if (r instanceof PossiblyNormalCompletion) {
+          throw ReportNoPossiblyNormalCompletionError();
+        }
+
+        result = GetValue(r, result, realm);
```
Comment (hermanventer) [inline src/evaluators/SwitchStatement.js]: Just use UpdateEmpty instead of GetValue.

File: src/evaluators/SwitchStatement.js [hunk 3]
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
...
+    }
+    // else we have a normal in-range case index
+
+    let c = cases[caseIndex];
+    let test = c.test;
+    invariant(test);
+
+    let selector = CaseSelectorEvaluation(test, strictCode, env, realm);
+    let selectionResult = computeBinary(realm, "===", input, selector);
+    if (selectionResult instanceof ConcreteValue) {
```
[the remark this answers is not in the document]
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Oh nice, I didn't realize!

File: src/evaluators/SwitchStatement.js [hunk 4]
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
...
+    // we start at the case we've been asked to evaluate, and process statements
+    // until there is either a break statement or exception thrown (this means we
+    // implicitly fall through correctly in the absence of a break statement).
+    while (caseIndex < cases.length && !(result instanceof Completion)) {
+      let c = cases[caseIndex];
+      for (let i = 0; i < c.consequent.length && !(result instanceof Completion); ++i) {
+        let node = c.consequent[i];
+        let r = env.evaluateCompletion(node, strictCode);
+
+        // TODO understand Reference and how to deal with it better
```
[the remark this answers is not in the document]
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Yay. That helps shed light on what Reference is. ;-)

File: src/evaluators/SwitchStatement.js [hunk 5]
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
...
+      if (To.ToBoolean(realm, selectionResult)) {
+        return DefiniteCaseEvaluation(caseIndex);
+      } else {
+        // we have a case that is definitively *not* taken
+        // so we go and look at the next one in the hope of finding a match
+        return AbstractCaseEvaluation(caseIndex + 1);
+      }
+    }
+
+    invariant(selectionResult instanceof AbstractValue);
```
[the remark this answers is not in the document]
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: yep!

File: src/evaluators/SwitchStatement.js [hunk 6]
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
...
+      for (let i = 0; i < c.consequent.length && !(result instanceof Completion); ++i) {
+        let node = c.consequent[i];
+        let r = env.evaluateCompletion(node, strictCode);
+
+        // TODO understand Reference and how to deal with it better
+        // TODO come up with a strategy for handling PossiblyNormalCompletion correctly
+        if (r instanceof Reference) {
+          throw ReportNoReferenceError();
+        } else if (r instanceof PossiblyNormalCompletion) {
+          throw ReportNoPossiblyNormalCompletionError();
```
[the remark this answers is not in the document]
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Of course. Note that in this case I am not making up exception classes. I am calling functions with those names that return the usual FatalError. I just wanted to wrap them up rather than repeat the code to report the diagnostic and throw the error each time I wanted to do it. So is using a wrapper function like this to do what you described a problem?

File: src/evaluators/SwitchStatement.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Makes sense, but Isn't that what happens? I think GetValue() makes result an AbruptCompletion in that case and then the test in the loop header will break out of the loop.

File: src/evaluators/SwitchStatement.js [hunk 7]
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
...
+
+    // TODO work out what to do with Reference
+    if (result instanceof Reference) {
+      throw ReportNoReferenceError();
+    }
+
+    if (result instanceof BreakCompletion) {
+      return result.value;
+    } else if (result instanceof AbruptCompletion) {
+      throw ReportNoAbruptCompletionError();
```
[the remark this answers is not in the document]
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Makes sense!

File: src/evaluators/SwitchStatement.js (hunk 1 above)
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Yay! Will remove it.

File: src/evaluators/SwitchStatement.js [hunk 8]
```diff
@@ -34,18 +36,170 @@ function CaseSelectorEvaluation(
...
+    } else if (result instanceof Completion) {
```
[the remark this answers is not in the document]
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: excellent. I can simplify this then.

File: src/evaluators/SwitchStatement.js (hunk 2 above)
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: ok.

File: src/evaluators/SwitchStatement.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (hermanventer) [reply src/evaluators/SwitchStatement.js]: It is easy to misread this pattern, which is what happened to me. The other reason you don't want to do this is that you end up with very different looking code when you decide to actually recover from an error.

File: src/evaluators/SwitchStatement.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (hermanventer) [reply src/evaluators/SwitchStatement.js]: Missed that. Putting the check in the loop header makes that logic non local and hard to spot when reading the code incrementally.

[the remark this answers is not in the document]
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Yep. I was trying to avoid having two if-then-break tests scattered around the loop, but...ugly as it is....it is probably much easier to understand. I'll change it.

File: src/evaluators/SwitchStatement.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Good point! Especially in Javascript where things are much more free flowing than my most recent implementation language (not good or bad...just different ;-)).

File: src/evaluators/SwitchStatement.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Interestingly, this then caused my tests to fail. It looks as though I have a little bit of work to do around handling Abrupt and PossiblyNormalCompletions correctly (beyond completions arising from break statements). So I've left it to error out of abstract evaluation if it sees a case that ends up in AbruptCompletion for now (just without the use of the helper function to do so). That way I can focus on finishing up this PR and addressing the completion issues (and i/o effects, expanding tests etc) in the next one.

**Comment by neilmacintosh** [comment]
I think I've addressed all the comments except the one about throwing AbruptCompletions. That exposed a bug in my handling of them which I'd like to postpone to the next PR.

File: src/evaluators/SwitchStatement.js [hunk 9]
```diff
@@ -90,32 +56,32 @@ function AbstractCaseBlockEvaluation(
...
+        invariant(!(r instanceof Reference));
 
-        // TODO understand Reference and how to deal with it better
         // TODO come up with a strategy for handling PossiblyNormalCompletion correctly
-        if (r instanceof Reference) {
-          throw ReportNoReferenceError();
-        } else if (r instanceof PossiblyNormalCompletion) {
-          throw ReportNoPossiblyNormalCompletionError();
+        if (r instanceof PossiblyNormalCompletion) {
+          let msg = "This operation is not yet supported on case blocks that result in a PossiblyNormalCompletion.";
```
Comment (hermanventer) [inline src/evaluators/SwitchStatement.js]: PossiblyNormalCompletion is an implementation detail that should not appear in a error message that is read by a developer using Prepack.

File: src/evaluators/SwitchStatement.js [hunk 10]
```diff
@@ -90,32 +56,32 @@ function AbstractCaseBlockEvaluation(
...
+          realm.reportIntrospectionError(msg);
```
Comment (hermanventer) [inline src/evaluators/SwitchStatement.js]: You want to create a CompilerDiagnostic, for which you'll need a new error code, which implies that you'll update the Wiki with a nice explanation of how such a situation arises and how to avoid it.

File: src/evaluators/SwitchStatement.js [hunk 11]
```diff
@@ -90,32 +56,32 @@ function AbstractCaseBlockEvaluation(
...
     }
 
     if (result instanceof BreakCompletion) {
       return result.value;
     } else if (result instanceof AbruptCompletion) {
-      throw ReportNoAbruptCompletionError();
-    } else if (result instanceof Completion) {
-      return result.value;
+      // TODO correct handling of PossiblyNormal and AbruptCompletion
+      let msg = "This operation is not yet supported on case blocks that throw.";
```
Comment (hermanventer) [inline src/evaluators/SwitchStatement.js]: Or return, or continue.

File: test/serializer/abstract/Switch.js [hunk 12]
```diff
@@ -61,12 +61,11 @@ switch (x) {
...
 }
 
-// throws introspection error
-switch (x) {
-  case 1: if (z10 === 13) z11 = 12; else throw 3; break;
-  case 0: throw 1; break;
-  case 2: z11 = 13; break;
-  default: throw 2; break;
-}
+// switch (x) {
```
Comment (hermanventer) [inline test/serializer/abstract/Switch.js]: Should this be commented out?

File: src/evaluators/SwitchStatement.js [hunk 13]
```diff
@@ -90,32 +56,32 @@ function AbstractCaseBlockEvaluation(
       for (let i = 0; i < c.consequent.length && !(result instanceof Completion); ++i) {
```
Comment (hermanventer) [inline src/evaluators/SwitchStatement.js]: !(result instanceof Completion) is always true here.

**Comment by hermanventer** [comment]
Time to update the PR title and summary.

File: test/serializer/abstract/Switch.js (hunk 12 above)
Comment (neilmacintosh) [reply test/serializer/abstract/Switch.js]: Ah! No. Nice catch.

File: src/evaluators/SwitchStatement.js (hunk 13 above)
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Ah yes...leftover from moving out the breaks. I'll kill it off, thanks.

File: src/evaluators/SwitchStatement.js (hunk 9 above)
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: ok. makes sense. I think your next comment has the punchline here ;-)

File: src/evaluators/SwitchStatement.js (hunk 10 above)
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Aha. I see now. I'll go look into how to do that and update this PR appropriately.

File: src/evaluators/SwitchStatement.js (hunk 11 above)
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Yep.

File: src/evaluators/SwitchStatement.js (hunk 10 above)
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Ok. Can you confirm that the correct place to describe the (hopefully temporary) limitation is here: https://prepack.io/frequently-asked-questions.html?

Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Ignore that, I found it!

Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: Ok. Created the page and uploading another diff to see if I have the error handling right.

File: src/evaluators/SwitchStatement.js [hunk 14]
```diff
@@ -36,16 +39,133 @@ function CaseSelectorEvaluation(
...
+    }
+  };
+
+  let AbstractCaseEvaluation = (caseIndex: number): Value => {
+    if (caseIndex === defaultCaseIndex) {
+      // skip the default case until we've exhausted all other options
+      return AbstractCaseEvaluation(caseIndex + 1);
+    } else if (caseIndex >= cases.length) {
+      // this is the stop condition for our recursive search for a matching case.
+      // we tried every available case index and if nothing matches we return
```
Comment (hermanventer) [inline src/evaluators/SwitchStatement.js]: and if -> and since

**Comment by neilmacintosh** [comment]
Ok. Updated title, summary and tag. Added proper emission of a compiler error (at least I think so) and a wiki page for the new error.

File: src/evaluators/SwitchStatement.js (hunk 14 above)
Comment (neilmacintosh) [reply src/evaluators/SwitchStatement.js]: thanks!
